### PR TITLE
chore(all): version trie calls

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

- [x] Version relevant trie calls (Put, Hash)
- [x] Convert wasmer Core_version `Version` to typed `trie.Version`
- [x] Adapt tests/test helpers to read the state version from the genesis runtime code
- [ ] Fix unit tests
- [ ] Fix integration tests

## Tests

Entire codebase test suite

## Issues

#2418 

## Primary Reviewer

@edwardmack

### CODE REVIEW

File: dot/core/service.go [hunk 1]
```diff
@@ -60,6 +61,7 @@ type Service struct {
 
 // Config holds the configuration for the core Service.
 type Config struct {
+	// TODO add state version field here
```
Comment (qdm12) [inline dot/core/service.go:64]: TODO

File: dot/rpc/modules/api.go [hunk 2]
```diff
@@ -101,7 +101,8 @@ type CoreAPI interface {
 	HandleSubmittedExtrinsic(types.Extrinsic) error
 	GetMetadata(bhash *common.Hash) ([]byte, error)
 	DecodeSessionKeys(enc []byte) ([]byte, error)
-	GetReadProofAt(block common.Hash, keys [][]byte) (common.Hash, [][]byte, error)
+	GetReadProofAt(block common.Hash, keys [][]byte) (
```
Comment (qdm12) [inline dot/rpc/modules/api.go:104]: missing state version arg?

File: dot/services.go [hunk 3]
```diff
@@ -121,7 +122,18 @@ func createRuntime(cfg *Config, ns runtime.NodeStorage, st *state.Service,
 		code = common.MustHexToBytes(codeString)
 	}
 
-	ts, err := st.Storage.TrieState(nil)
+	// Use the runtime state version to load the trie state in storage
```
Comment (qdm12) [inline dot/services.go:125]:
```suggestion
```

File: dot/services.go [hunk 4]
```diff
@@ -321,12 +333,24 @@ func (nodeBuilder) createRPCService(params rpcServiceSettings) (*rpc.HTTPServer,
...
+
 	genesisData, err := params.state.Base.LoadGenesisData()
 	if err != nil {
 		return nil, fmt.Errorf("failed to load genesis data: %s", err)
 	}
 
-	syncStateSrvc, err := modules.NewStateSync(genesisData, params.state.Storage)
+	// TODO should we inject the state version from the latest runtime?
+	// Or somehow find the genesis state version???
+	syncStateSrvc, err := modules.NewStateSync(genesisData, params.state.Storage, stateVersion)
```
Comment (qdm12) [inline dot/services.go:353]: TODO

File: dot/state/block.go [hunk 5]
```diff
@@ -651,6 +652,26 @@ func (bs *BlockState) GetRuntime(hash *common.Hash) (runtime.Instance, error) {
 	return bs.bt.GetBlockRuntime(*hash)
 }
 
+// GetRuntimeStateVersion returns the state version of the runtime
+// corresponding to the block hash given.
+func (bs *BlockState) GetRuntimeStateVersion(blockHash common.Hash) (
+	stateVersion trie.Version, err error) {
+	// TODO I don't think we should have instantiated instances in memory, we should just fetch
+	// the wasm code from disk and load it up when needed.
```
Comment (qdm12) [inline dot/state/block.go:660]:
```suggestion
```

File: dot/state/block.go [hunk 6]
```diff
@@ -651,6 +652,26 @@ func (bs *BlockState) GetRuntime(hash *common.Hash) (runtime.Instance, error) {
 	return bs.bt.GetBlockRuntime(*hash)
 }
 
+// GetRuntimeStateVersion returns the state version of the runtime
+// corresponding to the block hash given.
+func (bs *BlockState) GetRuntimeStateVersion(blockHash common.Hash) (
```
Comment (qdm12) [inline dot/state/block.go:657]: Use this all over the place

File: dot/state/offline_pruner.go [hunk 7]
```diff
@@ -118,7 +116,8 @@ func (p *OfflinePruner) SetBloomFilter() (err error) {
 	// loop from latest to last `retainBlockNum` blocks
 	for blockNum := header.Number; blockNum > 0 && blockNum >= latestBlockNum-uint(p.retainBlockNum); {
 		var tr *trie.Trie
-		tr, err = p.storageState.LoadFromDB(header.StateRoot)
+		// TODO do we need this pruner?
+		tr, err = p.storageState.LoadFromDB(header.StateRoot, trie.V0)
```
Comment (qdm12) [inline dot/state/offline_pruner.go:120]: TODO inject state version from somewhere? Maybe check out importAction

File: dot/state/offline_pruner.go [hunk 8]
```diff
@@ -118,7 +116,8 @@ func (p *OfflinePruner) SetBloomFilter() (err error) {
 	// loop from latest to last `retainBlockNum` blocks
 	for blockNum := header.Number; blockNum > 0 && blockNum >= latestBlockNum-uint(p.retainBlockNum); {
 		var tr *trie.Trie
-		tr, err = p.storageState.LoadFromDB(header.StateRoot)
+		// TODO do we need this pruner?
```
Comment (qdm12) [inline dot/state/offline_pruner.go:119]:
```suggestion
```

File: dot/core/service.go [hunk 9]
```diff
@@ -467,14 +475,21 @@ func (s *Service) GetRuntimeVersion(bhash *common.Hash) (
...
 
-	rt, err := s.blockState.GetRuntime(bhash)
+	stateVersion, err := rt.StateVersion()
 	if err != nil {
-		return version, err
+		return version, fmt.Errorf("getting runtime state version: %w", err)
+	}
+
+	// Note: not too sure why this trie state call is needed,
+	// but some RPC tests fail without it.
```
Comment (qdm12) [inline dot/core/service.go:489]: TODO see #2769

File: dot/state/block.go [hunk 10]
```diff
@@ -651,6 +652,23 @@ func (bs *BlockState) GetRuntime(hash *common.Hash) (runtime.Instance, error) {
...
+		return stateVersion, fmt.Errorf("getting runtime instance: %w", err)
+	}
+
+	stateVersion, err = instance.StateVersion()
+	if err != nil {
+		return stateVersion, fmt.Errorf("getting runtime state version: %w", err)
+	}
+
+	return stateVersion, nil
+}
```
Comment (qdm12) [inline dot/state/block.go:670]: Add unit/integ test

File: dot/sync/chain_processor.go [hunk 11]
```diff
@@ -196,32 +201,39 @@ func (s *chainProcessor) handleBody(body *types.Body) {
...
 	}
 
-	root := ts.MustRoot()
-	if !bytes.Equal(parent.StateRoot[:], root[:]) {
+	// TODO shall we remove this? Both are coming from the internal state
+	rootV0 := ts.MustRoot(trie.V0)
+	if !rootV0.Equal(parent.StateRoot) {
+		// TODO add extra check with rootV1 when v1 is supported
 		panic("parent state root does not match snapshot state root")
 	}
```
Comment (qdm12) [inline dot/sync/chain_processor.go:235]: Opinions on that? Should we remove it? Otherwise we'll have to hash tries for v0 and v1 which might have a performance impact.

File: lib/babe/babe.go [hunk 12]
```diff
@@ -489,9 +490,7 @@ func (b *Service) handleEpoch(epoch uint64) (next uint64, err error) {
 }
 
 func (b *Service) handleSlot(epoch, slotNum uint64,
-	authorityIndex uint32,
-	preRuntimeDigest *types.PreRuntimeDigest,
-) error {
+	authorityIndex uint32, preRuntimeDigest *types.PreRuntimeDigest) error {
```
Comment (qdm12) [inline lib/babe/babe.go:493]: revert

File: lib/babe/babe.go [hunk 13]
```diff
@@ -517,20 +516,22 @@ func (b *Service) handleSlot(epoch, slotNum uint64,
 	b.storageState.Lock()
 	defer b.storageState.Unlock()
 
+	hash := parent.Hash()
+	rt, err := b.blockState.GetRuntime(&hash)
+	if err != nil {
+		return fmt.Errorf("getting runtime: %w", err)
+	}
+
+	stateVersion := trie.Version(rt.Version().StateVersion)
```
Comment (qdm12) [inline lib/babe/babe.go:525]: use `rt.StateVersion()`

File: lib/babe/epoch_handler.go [hunk 14]
```diff
@@ -14,7 +14,8 @@ import (
 	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
 )
 
-type handleSlotFunc = func(epoch, slotNum uint64, authorityIndex uint32, preRuntimeDigest *types.PreRuntimeDigest) error
+type handleSlotFunc = func(epoch, slotNum uint64, authorityIndex uint32,
+	preRuntimeDigest *types.PreRuntimeDigest) error
```
Comment (qdm12) [inline lib/babe/epoch_handler.go:18]: revert

File: lib/babe/epoch_handler.go [hunk 15]
```diff
@@ -135,7 +136,8 @@ func (h *epochHandler) run(ctx context.Context, errCh chan<- error) {
 				panic(fmt.Sprintf("no VRF proof for authoring slot! slot=%d", swt.slotNum))
 			}
 
-			err := h.handleSlot(h.epochNumber, swt.slotNum, h.epochData.authorityIndex, h.slotToPreRuntimeDigest[swt.slotNum])
+			err := h.handleSlot(h.epochNumber, swt.slotNum, h.epochData.authorityIndex,
+				h.slotToPreRuntimeDigest[swt.slotNum])
```
Comment (qdm12) [inline lib/babe/epoch_handler.go:140]: revert

File: lib/genesis/genesis.go [hunk 16]
```diff
@@ -94,6 +97,23 @@ func (g *Genesis) ToRaw() error {
 	return nil
 }
 
+var ErrRuntimeCodeNotFound = errors.New("runtime code not found")
+
+func (g *Genesis) RuntimeCode() (runtimeCode []byte, err error) {
```
Comment (qdm12) [inline lib/genesis/genesis.go:102]: add unit test

File: lib/genesis/test_utils.go [hunk 17]
```diff
@@ -76,32 +76,32 @@ func CreateTestGenesisJSONFile(t *testing.T, fields Fields) (filename string) {
 }
 
 // NewTestGenesisWithTrieAndHeader generates genesis, genesis trie and genesis header
-func NewTestGenesisWithTrieAndHeader(t *testing.T) (*Genesis, *trie.Trie, *types.Header) {
+func NewTestGenesisWithTrieAndHeader(t *testing.T, version trie.Version) (*Genesis, *trie.Trie, *types.Header) {
```
Comment (qdm12) [inline lib/genesis/test_utils.go:79]: rename `version` to `stateVersion`

File: lib/genesis/test_utils.go [hunk 18]
```diff
@@ -76,32 +76,32 @@ func CreateTestGenesisJSONFile(t *testing.T, fields Fields) (filename string) {
...
 	require.NoError(t, err)
 
-	tr, h := newGenesisTrieAndHeader(t, gen)
+	tr, h := newGenesisTrieAndHeader(t, gen, version)
 	return gen, tr, h
 }
 
 // NewDevGenesisWithTrieAndHeader generates test dev genesis, genesis trie and genesis header
-func NewDevGenesisWithTrieAndHeader(t *testing.T) (*Genesis, *trie.Trie, *types.Header) {
+func NewDevGenesisWithTrieAndHeader(t *testing.T, version trie.Version) (*Genesis, *trie.Trie, *types.Header) {
```
Comment (qdm12) [inline lib/genesis/test_utils.go:89]: rename `version` to `stateVersion`

File: lib/genesis/test_utils.go [hunk 19]
```diff
@@ -76,32 +76,32 @@ func CreateTestGenesisJSONFile(t *testing.T, fields Fields) (filename string) {
...
 	require.NoError(t, err)
 
-	tr, h := newGenesisTrieAndHeader(t, gen)
+	tr, h := newGenesisTrieAndHeader(t, gen, version)
 	return gen, tr, h
 }
 
-func newGenesisTrieAndHeader(t *testing.T, gen *Genesis) (*trie.Trie, *types.Header) {
-	genTrie, err := NewTrieFromGenesis(gen)
+func newGenesisTrieAndHeader(t *testing.T, gen *Genesis, version trie.Version) (*trie.Trie, *types.Header) {
```
Comment (qdm12) [inline lib/genesis/test_utils.go:99]: rename `version` to `stateVersion`

File: lib/runtime/allocator.go [hunk 20]
```diff
@@ -82,9 +82,11 @@ func (fbha *FreeingBumpHeapAllocator) growHeap(numPages uint32) error {
 	return nil
 }
 
-// Allocate determines if there is space available in WASM heap to grow the heap by 'size'.  If there is space
-//   available it grows the heap to fit give 'size'.  The heap grows is chunks of Powers of 2, so the growth becomes
-//   the next highest power of 2 of the requested size.
+// Allocate determines if there is space available in WASM heap to grow
+// the heap by 'size' bytes. If there is space available it grows the heap
+//  to fit the given 'size' number of bytes. The heap grows is chunks of
```
Comment (qdm12) [inline lib/runtime/allocator.go:87]:
```suggestion
// to fit the given 'size' number of bytes. The heap grows is chunks of
```

and move to another pr

File: lib/runtime/interface.go [hunk 21]
```diff
@@ -50,11 +51,11 @@ type Instance interface {
 
 // Storage interface
 type Storage interface {
-	Set(key []byte, value []byte)
+	Set(key []byte, value []byte, version trie.Version)
```
Comment (qdm12) [inline lib/runtime/interface.go:54]: rename `version` to `stateVersion`

File: lib/runtime/interface.go [hunk 22]
```diff
@@ -50,11 +51,11 @@ type Instance interface {
...
 	Get(key []byte) []byte
-	Root() (common.Hash, error)
-	SetChild(keyToChild []byte, child *trie.Trie) error
-	SetChildStorage(keyToChild, key, value []byte) error
+	Root(version trie.Version) (common.Hash, error)
+	SetChild(keyToChild []byte, child *trie.Trie, version trie.Version) error
+	SetChildStorage(keyToChild, key, value []byte, version trie.Version) error
```
Comment (qdm12) [inline lib/runtime/interface.go:58]: rename `version` to `stateVersion`

File: lib/runtime/wasmer/exports.go [hunk 23]
```diff
@@ -53,6 +54,13 @@ func (in *Instance) version() (version runtime.Version, err error) {
 	return version, nil
 }
 
+// StateVersion returns the typed instance state version.
+// This is cheap to call since the instance version is cached.
+func (in *Instance) StateVersion() (stateVersion trie.Version, err error) {
+	coreVersion := in.ctx.Version
+	return trie.ParseVersion(coreVersion.StateVersion)
```
Comment (qdm12) [inline lib/runtime/wasmer/exports.go:61]: add unit test

File: lib/runtime/wasmer/imports.go [hunk 24]
```diff
@@ -1173,7 +1173,8 @@ func ext_default_child_storage_set_version_1(context unsafe.Pointer,
 	cp := make([]byte, len(value))
 	copy(cp, value)
 
-	err := storage.SetChildStorage(childStorageKey, key, cp)
+	// TODO obtain version from context
+	err := storage.SetChildStorage(childStorageKey, key, cp, trie.V0)
```
Comment (qdm12) [inline lib/runtime/wasmer/imports.go:1177]: TODO!

File: lib/runtime/wasmer/imports.go [hunk 25]
```diff
@@ -1142,7 +1142,7 @@ func ext_default_child_storage_root_version_1(context unsafe.Pointer,
 		return 0
 	}
 
-	childRoot, err := child.Hash()
+	childRoot, err := child.Hash(trie.V0)
```
Comment (qdm12) [inline lib/runtime/wasmer/imports.go:1145]: Get from instance context? Check v1 state trie runtime functions

File: lib/trie/trie.go [hunk 26]
```diff
@@ -13,8 +13,11 @@ import (
 	"github.com/ChainSafe/gossamer/lib/common"
 )
 
-// EmptyHash is the empty trie hash.
-var EmptyHash, _ = NewEmptyTrie().Hash()
+var (
+	// EmptyHash is the empty trie hash, which is the same for
+	// both the V0 and V1 state trie versions
+	EmptyHash, _ = NewEmptyTrie().Hash(V0)
+)
```
Comment (qdm12) [inline lib/trie/trie.go:20]: nit Remove parentheses

File: dot/services.go (hunk 4 above)
Comment (qdm12) [reply dot/services.go:342]: From Kishan

> I feel like it would be best block runtime
> substrate has features to update the runtime. So, with that in mind taking info from best block might make more sense.

From Eclesio

> Given the current substrate implementation you should use the latest finalized block -> https://github.com/paritytech/substrate/blob/eddf888394cba28b048bd79e319e8b27f4b414f3/client/sync-state-rpc/src/lib.rs#L142
> So you can use the finalized block runtime

**Comment by dimartiro** [comment]
Closed in favor of #3433
